README.md
# uvc-util
USB Video Class (UVC) control management utility for Mac OS X

This code arose from a need for a command-line utility on Mac OS X that could query and modify UVC camera controls (like contrast and brightness).  It presently implements all Terminal and Processing Unit controls available under the [1.1 standard](http://www.cajunbot.com/wiki/images/8/85/USB_Video_Class_1.1.pdf "UVC 1.1 PDF").  Additional [1.5 standard](https://www.usb.org/sites/default/files/USB_Video_Class_1_5.zip) Terminal and Processing Unit controls were added for the 1.2 release of this software.

Control values are implemented using a class (UVCType) that represents byte-packed data structures containing core atomic types (8-, 16-, 32-, and 64-bit integers).  Multi-component types allow fields to be named.  Another class (UVCValue) uses UVCType and a memory buffer to manage data structured according to that UVCType.  Thus, the code knows how each implemented UVC control's data is structured, which allows for per-component byte-swapping when necessary, etc.

Unlike other (GUI-based) utilities, this code only makes use of the IOKit to walk the USB bus, searching for UVC-compliant devices.

The utility's help screen:

~~~~
usage:

    ./uvc-util {options/actions/target selection}

  Options:

    -h/--help                              Show this information
    -v/--version                           Show the version of the program
    -k/--keep-running                      Continue processing additional actions despite
                                           encountering errors

  Actions:

    -d/--list-devices                      Display a list of all UVC-capable devices
    -c/--list-controls                     Display a list of UVC controls implemented

    Available after a target device is selected:

    -c/--list-controls                     Display a list of UVC controls available for
                                           the target device

    -S <control-name>                      Display available information for the given
    --show-control=<control-name>          UVC control:  component fields for multi-value
                                           types, minimum, maximum, resolution, and default
                                           value when provided:

        pan-tilt-abs {
          type-description: {
            signed 32-bit integer            pan;
            signed 32-bit integer            tilt;
          },
          minimum: {pan=-648000,tilt=-648000}
          maximum: {pan=648000,tilt=648000}
          step-size: {pan=3600,tilt=3600}
          default-value: {pan=0,tilt=0}
        }

    -g <control-name>                      Get the value of a control.
    --get=<control-name>

    -o <control-name>                      Same as -g/--get, but ONLY the value of the control
    --get-value=<control-name>             is displayed (no label)

    -s <control-name>=<value>              Set the value of a control; see below for a
    --set=<control-name>=<value>           description of <value>

    Specifying <value> for -s/--set:

      * The string "default" indicates the control should be reset to its default value(s)
        (if available)
      * The string "minimum" indicates the control should be reset to its minimum value(s)
        (if available)
      * The string "maximum" indicates the control should be reset to its maximum value(s)
        (if available)

      * Multi-component controls must provide a list of per-component values.  The values may
        be specified either in the same sequence as shown by the -S/--show-control, or by naming
        each value.  For example, the "pan-tilt-abs" control has two components, "pan" and
        "tilt" (in that order), so the following are equivalent:

            -s pan-tilt-abs="{-3600, 36000}"
            -s pan-tilt-abs="{tilt=0.52778, pan=-3600}"

      * Single-value controls should not use the brace notation, just the component value of the
        control, for example:

            -s brightness=0.5

      * Component values may be provided as fractional values (in the range [0,1]) if the control
        provides a value range (can be checked using -S/--show-control).  The value "0.0"
        corresponds to the minimum, "1.0" to the maximum.

      * Component values may use the strings "default," "minimum," or "maximum" to indicate that
        the component's default, minimum, or maximum value should be used (if the control provides one,
        can be checked using -S/--show-control)

            -s pan-tilt-abs="{default,minimum}"
            -s pan-tilt-abs="{tilt=-648000,pan=default}"

  Methods for selecting the target device:

    -0
    --select-none

         Drop the selected target device

    -I <device-index>
    --select-by-index=<device-index>

         Index of the device in the list of all devices (zero-based)

    -V <vendor-id>:<product-id>
    --select-by-vendor-and-product-id=<vendor-id>:<product-id>

         Provide the hexadecimal- or integer-valued vendor and product identifier
         (Prefix hexadecimal values with "0x")

    -L <location-id>
    --select-by-location-id=<location-id>

         Provide the hexadecimal- or integer-valued USB locationID attribute
         (Prefix hexadecimal values with "0x")

    -N <device-name>
    --select-by-name=<device-name>

         Provide the USB product name (e.g. "AV.io HDMI Video")

~~~~



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0]
Baseline release to open source.

## [1.2.0] - 2021-04-25
### Fixed
- Several issues reported with some camera's ProcessingUnit being available but unusable.  Diagnosed as the use of a static unit id of 2 for the ProcessingUnit, whereas the UVC standard has a variable unit id present in the PU header.  Added unit id map to UVCController with default unit ids for each handled unit type, overridden by unit id from the unit's header record.  User who reported issues tested the change, confirmed success.


src/UVCType.h
//
//  UVCType.h
//
//  Abstract data types for UVC controls.
//
//  Copyright © 2016
//  Dr. Jeffrey Frey, IT-NSS
//  University of Delaware
//
// $Id$
//

#import <Foundation/Foundation.h>

/*!
  @typedef UVCTypeComponentType
  
  Enumerates the atomic data types that the UVCType class implements.
  These are underlying types used by the UVC standard in control
  interfaces.
*/
typedef enum {
  kUVCTypeComponentTypeInvalid   = 0,
  kUVCTypeComponentTypeBoolean,
  kUVCTypeComponentTypeSInt8,
  kUVCTypeComponentTypeUInt8,
  kUVCTypeComponentTypeBitmap8,
  kUVCTypeComponentTypeSInt16,
  kUVCTypeComponentTypeUInt16,
  kUVCTypeComponentTypeSInt32,
  kUVCTypeComponentTypeUInt32,
  kUVCTypeComponentTypeSInt64,
  kUVCTypeComponentTypeUInt64,
  kUVCTypeComponentTypeMax
} UVCTypeComponentType;

/*!
  @function UVCTypeComponentByteSize
  
  Returns the number of bytes occupied by the given componentType or
  zero (0) if componentType was invalid.
*/
NSUInteger UVCTypeComponentByteSize(UVCTypeComponentType componentType);

/*!
  @defined UVCTypeInvalidIndex
  
  Constant returned by UVCType to indicate that a field index was out
  of range.
*/
#define UVCTypeInvalidIndex NSUIntegerMax

/*!
  @typedef UVCTypeScanFlags
  
  Enumerates bitmask components that alter the behavior of the scanCString:*
  methods of UVCType.
  
  The kUVCTypeScanFlagShowWarnings flag allows warning messages to be
  written to stderr as the routines parse a cString.  Additionallly, the
  kUVCTypeScanFlagShowInfo flag produces more extensive output to stderr
  as the cString is processed (more like debugging information). 
*/
typedef enum {
  kUVCTypeScanFlagShowWarnings    = 1 << 0,
  kUVCTypeScanFlagShowInfo        = 1 << 1
} UVCTypeScanFlags;

/*!
  @class UVCType
  @abstract Abstract data type comprised of UVCTypeComponentType atomic types
  
  Instances of the UVCType class represent the structured data brokered by
  UVC controls.  A UVCType comprises one or more named data structure fields in
  a specific order, with each having an atomic type from the
  UVCTypeComponentType enumeration.
  
  The set of fields correlate to a packed C struct without word boundary padding;
  this also correlates directly to the format of UVC control data.
  
  In case this code were to be compiled on a big-endian host, byte-swapping
  routines are included which can reorder an external buffer (containing the
  UVC control data structured by the UVCType) to and from USB (little) endian.
  
  Methods are included to calculate relative byte offsets of the component fields,
  either by index of the name of the field.  Also, the number of bytes occupied
  by the UVCType is available via the byteSize method.
  
  Methods are also provided to initialize an external buffer structured by a
  UVCType using textual input (from a C string).
*/
@interface UVCType : NSObject
{
  NSUInteger    _fieldCount;
  BOOL          _needsNoByteSwap;
  void          *_fields;
}

/*!
  @method uvcTypeWithCString:
  
  Returns an autoreleased instance of UVCType initialized with the component field(s)
  described by a C string.  The C string must begin and end with curly braces and
  include one or more named types.  Each named type follows the syntax:
  
    [type] [name];
    
  where [name] is the component name (alphanumeric characters and '-') and the [type]
  is one of:
  
    [type]    description
    B         UInt8, accepting only 0 and 1
    S1        SInt8 / char
    U1        UInt8 / unsigned char
    S2        SInt16 / short
    U2        UInt16 / unsigned short
    S4        SInt32 / int
    U4        UInt32 / unsigned int
    S8        SInt64 / long long int
    U8        UInt64 / unsigned long long int
  
  For types with a single field, the [name] can be omitted:
  
    {S2}
    
  Other examples:
  
    { S2 pan; S2 tilt; }
    
*/
+ (UVCType*) uvcTypeWithCString:(const char*)typeDescription;

/*!
  @method uvcTypeWithFieldNamesAndTypes:
  
  Returns an autoreleased instance of UVCType initialized with an arbitrary length sequence
  of component field names and types, followed by nil or NULL.  For example:
  
    [uvcTypeWithFieldNamesAndTypes:@"pan", kUVCTypeComponentTypeSInt16, @"tilt", kUVCTypeComponentTypeSInt16, nil];

*/
+ (UVCType*) uvcTypeWithFieldNamesAndTypes:(NSString*)name,...;

/*!
  @method uvcTypeWithFieldCount:names:types:
  
  Returns an autoreleased instance of UVCType initialized using two C arrays of component
  field names (as NSString instances) and types.  The arrays must have at least count elements
  present.
*/
+ (UVCType*) uvcTypeWithFieldCount:(NSUInteger)count names:(NSString**)names types:(UVCTypeComponentType*)types;

/*!
  @method fieldCount
  
  Returns the number of component fields in the structure represented by the receiver.
*/
- (NSUInteger) fieldCount;

/*!
  @method fieldNameAtIndex:
  
  Returns the name associated with the component field at the given index.
  
  Returns nil if index is out of range.
*/
- (NSString*) fieldNameAtIndex:(NSUInteger)index;

/*!
  @method fieldTypeAtIndex:
  
  Returns the type associated with the component field at the given index.
  
  Returns kUVCTypeComponentTypeInvalid if index is out of range.
*/
- (UVCTypeComponentType) fieldTypeAtIndex:(NSUInteger)index;

/*!
  @method indexOfFieldWithName:
  
  If one of the receiver's component fields is named the same as fieldName (under a case-insensitive
  string comparison) returns the index of that field.  Otherwise, UVCTypeInvalidIndex is returned.
*/
- (NSUInteger) indexOfFieldWithName:(NSString*)fieldName;

/*!
  @method byteSize
  
  Returns the number of bytes that data structured according to the receiver's component field
  types would occupy.
*/
- (NSUInteger) byteSize;

/*!
  @method offsetToFieldAtIndex:
  
  Returns the relative offset (in bytes) at which the given component field would be found in a
  buffer structured according to the receiver's component field types.
  
  Returns UVCTypeInvalidIndex if index is out of range.
*/
- (NSUInteger) offsetToFieldAtIndex:(NSUInteger)index;

/*!
  @method offsetToFieldWithName:
  
  Returns the relative offset (in bytes) at which the given component field (identified by
  case-insensitive string comparison against fieldName) would be found in a buffer structured
  according to the receiver's component field types.
  
  Returns UVCTypeInvalidIndex if index is out of range.
*/
- (NSUInteger) offsetToFieldWithName:(NSString*)fieldName;

/*!
  @method byteSwapHostToUSBEndian:
  
  Given an external buffer structured according to the receiver's component field types,
  byte swap all necessary component fields (anything larger than 1 byte) from the host endian
  to USB (little) endian.
*/
- (void) byteSwapHostToUSBEndian:(void*)buffer;

/*!
  @method byteSwapUSBToHostEndian:
  
  Given an external buffer structured according to the receiver's component field types,
  byte swap all necessary component fields (anything larger than 1 byte) from USB (little)
  endian to host endian.
*/
- (void) byteSwapUSBToHostEndian:(void*)buffer;

/*!
  @method scanCString:intoBuffer:flags:
  
  Convenience method that calls
  
    [self scanCString:cString intoBuffer:buffer flags:flags minimum:NULL maximum:NULL stepSize:NULL defaultValue:NULL]
*/
- (BOOL) scanCString:(const char*)cString intoBuffer:(void*)buffer flags:(UVCTypeScanFlags)flags;

/*!
  @method scanCString:intoBuffer:flags:minimum:maximum:
  
  Convenience method that calls
  
    [self scanCString:cString intoBuffer:buffer flags:flags minimum:minimum maximum:maximum stepSize:NULL defaultValue:NULL]
*/
- (BOOL) scanCString:(const char*)cString intoBuffer:(void*)buffer flags:(UVCTypeScanFlags)flags minimum:(void*)minimum maximum:(void*)maximum;

/*!
  @method scanCString:intoBuffer:flags:minimum:maximum:stepSize:
  
  Convenience method that calls
  
    [self scanCString:cString intoBuffer:buffer flags:flags minimum:minimum maximum:maximum stepSize:stepSize defaultValue:NULL]
*/
- (BOOL) scanCString:(const char*)cString intoBuffer:(void*)buffer flags:(UVCTypeScanFlags)flags minimum:(void*)minimum maximum:(void*)maximum stepSize:(void*)stepSize;

/*!
  @method scanCString:intoBuffer:flags:minimum:maximum:stepSize:defaultValue:
  
  The arguments buffer, minimum, maximum, stepSize, and defaultValue are all assumed to be at least as
  large as the receiver's [self byteSize] method would return.  If any of minimum, maximum, stepSize, or
  defaultValue are non-NULL, they are assumed to contain the corresponding limit data read from the UVC
  device.
  
  Parses cString and attempts to fill-in the provided buffer according to the receiver's component fields.
  The component values must be contained within curly braces, and values should be delimited using a
  comma.  Whitespace is permissible around words and commas.
  
  Use of a floating-point (fractional) value for a component requires that minimum and maximum are non-NULL.
  The fractional value maps to the correponding value in that range, with 0.0f being the minimum.
  
  The words "default," "minimum," and "maximum" are permissible on components so long as minimum,
  maximum, or defaultValue are non-NULL.
  
  Receivers with a single component may omit the curly braces.
  
  If the words "default," "minimum," or "maximum" are the entirety of cString, then the corresponding
  value is set for all component fields, provided the corresponding pointer argument is non-NULL.
  
  The flags bitmask controls this function's behavior.  At this time, the only flags present enable
  informative output to stderr as the parsing progresses.
  
  Returns YES if all component fields of the receiver were successfully set in the provided buffer.
  
  Examples:
    
    Type            Value
    {S2}            "25"
    {S2}            "{=25}" (same as previous)
    {S2 x; S2 y;}   "{0,1}"
    {S2 x; S2 y;}   "{y=1,x=0}" (same as previous)
    {S2 w; S2 h;}   "{h=minimum,w=default}" (provided minimum and defaultValue are non-NULL)
    {S2 w; S2 h;}   "{default,minimum}" (same as previous)

*/
- (BOOL) scanCString:(const char*)cString intoBuffer:(void*)buffer flags:(UVCTypeScanFlags)flags minimum:(void*)minimum maximum:(void*)maximum stepSize:(void*)stepSize defaultValue:(void*)defaultValue;

/*!
  @method stringFromBuffer:
  
  Create a formatted textual description of the data in the external buffer structured according to
  the receiver's component field types.  E.g.
  
    "{pan=3600,tilt=-360000}"
    
*/
- (NSString*) stringFromBuffer:(void*)buffer;

/*!
  @method typeSummaryString
  
  Returns a human-readable description of the receiver's component field types.
*/
- (NSString*) typeSummaryString;

@end
